libraries/lib-snapping/SnapUtils.cpp

```diff
@@ -0,0 +1,315 @@
+
+   it = cache.find(id);
+
+   return it != cache.end() ? it->second : nullptr;
+}
+
+SnapResult SnapFunctionsRegistry::Snap(
+   const Identifier& id, SnapConfig config, double time, bool nearest)
+{
+   auto item = Find(id);
+
+   if (item == nullptr)
+      return SnapResult { time, false };
+
+   return item->snapFunction(config, time, nearest);
+}
+
+SnapRegistryGroup::~SnapRegistryGroup()
+{
+}
+
+bool SnapRegistryGroup::Transparent() const
+{
+   return transparent;
+}
+
+SnapRegistryItem::SnapRegistryItem(
+   const Identifier& internalName, const TranslatableString& _label,
+   SnapFunctor _snapFunction)
+    : SingleItem { internalName }
+    , label { _label }
+    , snapFunction { _snapFunction }
+{
+}
+
+SnapRegistryItem::~SnapRegistryItem()
+{
+}
+
+Registry::BaseItemPtr SnapFunction(
+   const Identifier& functionId, const TranslatableString& label,
+   SnapFunctor functor)
+{
+   return std::make_unique<SnapRegistryItem>(
+      functionId, label, std::move(functor));
+}
+
+SnapRegistryItemRegistrator::SnapRegistryItemRegistrator(
+   const Registry::Placement& placement, Registry::BaseItemPtr pItem)
+    : RegisteredItem { std::move(pItem), placement }
+{
+}
+
+namespace
+{
+SnapResult SnapWithMultiplier (double value, double multiplier, bool nearest)
+{
+   auto result = nearest ? std::round(value * multiplier) / multiplier :
+                           std::floor(value * multiplier) / multiplier;
+
+   return SnapResult { result, true };
+}
+
+
+SnapFunctor SnapToTime(double multiplier) {
+   return [multiplier](SnapConfig, double value, bool nearest)
+   { return SnapWithMultiplier(value, multiplier, nearest); };
+}
+
+SnapResult SnapToSamples (SnapConfig cfg, double value, bool nearest)
+{
+   return SnapWithMultiplier(value, cfg.rate, nearest);
+}
+
+SnapFunctor SnapToFrames(double fps)
+{
+   return [fps](SnapConfig, double value, bool nearest)
+   { return SnapWithMultiplier(value, fps, nearest); };
+}
+
+/*
+   bps = tempo / 60
+   1/lower takes 1/bps
+*/
+
+SnapFunctor SnapToBar ()
+{
+   return [](SnapConfig cfg, double value, bool nearest)
+   {
+      const auto multiplier = (cfg.tempo / 60.0) / cfg.timeSignature.first;
+      return SnapWithMultiplier(value, multiplier, nearest);
+   };
+}
+
+SnapFunctor SnapToBeat(int divisor)
+{
+   return [divisor](SnapConfig cfg, double value, bool nearest)
+   {
+      const auto multiplier =
+         (cfg.tempo / 60.0) * divisor / cfg.timeSignature.second;
+      return SnapWithMultiplier(value, multiplier, nearest);
+   };
+}
+
+SnapFunctor SnapToTriplets(int divisor)
+{
+   return [divisor](SnapConfig cfg, double value, bool nearest)
+   {
+      const auto tripletDivisor = 3 * (divisor / 2);
+      const auto multiplier =
+         (cfg.tempo / 60.0) * tripletDivisor / cfg.timeSignature.second;
+      return SnapWithMultiplier(value, multiplier, nearest);
+   };
+}
+
+SnapRegistryItemRegistrator beats {
+   Registry::Placement { {}, { Registry::OrderingHint::Begin } },
```

Comment:
The ordering preference initializer makes hints unnecessary.

Hints will be useful if ever there are scattered snap function definitions in other files

Reply:
I would love to omit the ordering preference initializer, but turns out nothing works without it. By the time I figured that out I've spent to much time already to continue experiment on how things work. 

I would like to keep it this way though, this improves the readability somewhat.

Reply:
But too I think the functions might be scattered to other files in a later library reorganization (which I don't demand with this PR).  The snapping routines special to beats and also the numeric format special to beats would live together and push their special case stuff into registries.

Reply:
I still agree it ought to be easier to write and it’s on me to improve Registry.h so you can

Reply:
It should look more like a menu definition — which uses things not all in Registry.h.  There are template patterns there that could become more general utilities.